docs/code/MlNetCookBook.md

```diff
@@ -0,0 +1,707 @@
+- MeanVar normalizer: for each feature, compute the mean and variance, and then linearly rescale it to zero-mean, unit-variance.
+- CDF normalizer: for each feature, compute the mean and variance, and then replace each value `x` with `Cdf(x)`, where `Cdf` is the cumulative density function of normal distribution with these mean and variance. 
+- Binning normalizer: discretize the feature value into `N` 'buckets', and then replace each value with the index of the bucket, divided by `N-1`.
+
+These normalizers all have different properties and tradeoffs, but it's not *that* big of a deal if you use one over another. Just make sure you use a normalizer when training linear models or other parametric models. 
+
+An important parameter of ML.NET normalizers is called `fixZero`. If `fixZero` is true, zero input is always mapped to zero output. This is very important when you handle sparse data: if we don't preserve zeroes, we will turn all sparse data into dense, which is usually a bad idea.
+
+It is a good practice to include the normalizer directly in the ML.NET learning pipeline: this way you are sure that the normalization
+- is only trained on the training data, and not on your test data,
+- is correctly applied to all the new incoming data, without the need for extra pre-processing at prediction time.
+
+Here's a snippet of code that demonstrates normalization in learning pipelines. It assumes the Iris dataset:
+```c#
+// Create a new environment for ML.NET operations. It can be used for exception tracking and logging, 
+// as well as the source of randomness.
+var env = new LocalEnvironment();
+
+// Define the reader: specify the data columns and where to find them in the text file.
+var reader = TextLoader.CreateReader(env, ctx => (
+        // The four features of the Iris dataset will be grouped together as one Features column.
+        Features: ctx.LoadFloat(0, 3),
+        // Label: kind of iris.
+        Label: ctx.LoadText(4)
+    ),
+    // Default separator is tab, but the dataset has comma.
+    separator: ',');
+
+// Read the training data.
+var trainData = reader.Read(new MultiFileSource(dataPath));
+
+// Apply all kinds of standard ML.NET normalization to the raw features.
+var pipeline = reader.MakeNewEstimator()
+    .Append(r => (
+        MinMaxNormalized: r.Features.Normalize(fixZero: true),
+        MeanVarNormalized: r.Features.NormalizeByMeanVar(fixZero: false),
+        CdfNormalized: r.Features.NormalizeByCumulativeDistribution(),
+        BinNormalized: r.Features.NormalizeByBinning(maxBins: 256),
+    ));
+
+// Let's train our pipeline of normalizers, and then apply it to the same data.
+var normalizedData = pipeline.Fit(trainData).Transform(trainData);
+
+// Inspect one column of the resulting dataset.
+var meanVarValues = normalizedData.GetColumn(r => r.MeanVarNormalized).ToArray();
+```
```

Comment:
Is it worth showing how to get the normalization weights? This is a commonly asked question. Alternatively, generalizing the idea of onFit from the previous section to explain that it can be used in other components.

Reply:
What do you think about including this?

Reply:
I will

---
In reply to: [221339370](https://github.com/dotnet/machinelearning/pull/960#discussion_r221339370) [](ancestors = 221339370)